packages/server-wallet/src/protocols/process-ledger-queue.ts

```diff
@@ -1,11 +1,36 @@
+/**
+ * Follows an algorithm to guarantee that an agreed upon ledger update will be signed by
+ * 2 participants in at most 2 round trips. Either can propose a state at any time and either
+ * can counterpropose a state at any time. However, if they follow the algorithm below they
+ * should always be able to compute the expected outcome by the second round trip.
+ *
+ * Algorithm:
+ *
+ * IF there are any pending ledger updates (to fund OR defund a channel)
+ *
+ *   i. Check if I already proposed a new ledger update O₁. Otherwise, iteratively allocate funds
+ *      from the supported state's (i.e., S) outcome to each request, call this O₁.
+ *
+ *  ii. Check if there is an existing proposal for a ledger update, O₂.
+ *
+ * iii. Compute O₁ ⋂ O₂ (NOTE: O₂ might be NULL i.e., not received yet)
+ *
+ *      a. If I haven't sent my proposal yet, propose O₁ ⋂ O₂.
+ *
+ *      b. If I have sent my proposal and received theirs sign O₁ ⋂ O₂ at
+ *         the turn number 2 higher than the most recently supported one.
+ *
+ *      c. If O₂ is NULL (not received a proposal), wait for the counterparty.
```

Comment:
This allows for a deadlock:
1. my message with my proposal got dropped
2. their message with their proposal got dropped

Reply:
Started a related discussion here: https://github.com/statechannels/statechannels/issues/3057

One strategy could be to automatically trigger a channel sync if this stage of the protocol is reached. I'm not opposed to doing that.

Reply:
> Started a related discussion here: #3057

👍 

I think we need some tighter organization of this project. Creating issues is good, but we ought to have a "checklist" of things that need to get done to ship a robust ledger funding protocol.

> One strategy could be to automatically trigger a channel sync if this stage of the protocol is reached. I'm not opposed to doing that.

I think this is the most natural solution here. It is a bit complex, because you want to ensure you don't enter an infinite loop in the case where channel syncing fails to change state on either end.

Reply:
> I think we need some tighter organization of this project. Creating issues is good, but we ought to have a "checklist" of things that need to get done to ship a robust ledger funding protocol.
> 

I've created some labels for challenging-v0 to tag issues so maybe the same thing could work here?

Reply:
I like that suggestion, and I'd be keen to give that process a try for one or two projects.